MMseqs2 1-c7a89 (new formula)

milot-mirdita opened the pull request:
Mseqs2 is a bioinformatics software suite for
ultra fast and sensitive protein sequence
searches and clustering.

I also tested this recipe with linuxbrew
on Ubuntu 16.04.

Originally, I was planning on submitting this to
homebrew/science, but it seems to be in the process
of being phased out.

There are a some potential issues with this recipe:
* Zlib/bzip2 dependency, instead of system ones
  The system zlib errors out with something like
  __OSX_AVAILABLE_STARTING 10_13 on my OSX Sierra.
  bzip2 works, but for consistencies sake, i added
  a dependecy for both.
  brew audit is complaining about this issue
* CPU instructions set is per default autodetected.
  At least SSE4.1 is required for MMseqs2 to work.
  The --with-sse41/avx2 flags force a certain compilation.
  Bottles are always built with sse41 as the lowest
  common denominator. A 'depends_on :cpuid => :sse41',
  would be useful to handle this cleanly.
  I added a caveat about this to the recipe.
* The MMseqs2 build process either needs xxd from vim
  or perl (to run the bundled xxdi.pl script).
  I didn't include either as an dependecy, since
  pretty much every OS (also linuxbrew)  include perl.
  Should I explicitly add perl or vim as a build
  dependency? I would like to keep dependencies to
  a minimum.

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

Not quite, I have explicitly added zlib and bzip2 as dependencies, since I cannot successfully use the system zlib on MacOS Sierra with the newest Xcode (see above).

-----

martin-steinegger commented:
@ilovezfs is there anything missing to merge this pull request?

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,71 @@
+class Mmseqs2 < Formula
+  desc "Software suite for very fast protein sequence search and clustering"
+  homepage "https://mmseqs.org"
+  # tag "bioinformatics"
```
No `tag` and `doi`

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,71 @@
+  homepage "https://mmseqs.org"
+  # tag "bioinformatics"
+  # doi "10.1038/nbt.3988"
+  url "https://github.com/soedinglab/MMseqs2.git", :using => :git, :tag => "1-c7a89", :revision => "c7a89b6aa6d94796783e94d49750b3e0882573aa"
```
We prefer to build from release tarballs, like https://github.com/soedinglab/MMseqs2/archive/1-c7a89.tar.gz

milot-mirdita replied:
If I use the release tarball, MMseqs2 will not disable the git commit used to compile itself.

I can include a hack like this in the formula:
`inreplace "src/version/Version.cpp", /.+/m, "const char *version = \"#{version}\";"`
and fix it in some cleaner way for the next release. Would that be okay?

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,71 @@
+  url "https://github.com/soedinglab/MMseqs2.git", :using => :git, :tag => "1-c7a89", :revision => "c7a89b6aa6d94796783e94d49750b3e0882573aa"
+  version "1-c7a89"
+  head do
+    url "https://github.com/soedinglab/MMseqs2.git", :using => :git
```
`, :using => :git` is not needed

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,71 @@
+  head do
+    url "https://github.com/soedinglab/MMseqs2.git", :using => :git
+    resource "documentation" do
+      url "https://github.com/soedinglab/MMseqs2.wiki.git", :using => :git
```
`, :using => :git` is not needed

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,71 @@
+  depends_on :arch => :x86_64
+
+  resource "documentation" do
+    url "https://github.com/soedinglab/MMseqs2.wiki.git", :using => :git, :revision => "6dbd3666edb64fc71173ee714014e88c1ebe2dfc"
```
`, :using => :git` is not needed

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,71 @@
+  end
+
+  option "with-avx2", "Build MMseqs2 with AVX2 instruction set support"
+  option "with-sse41", "Build MMseqs2 with SSE4.1 instruction set support"
```
If software doesn't build without SSE4.1 support, don't make it an option.

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,71 @@
+    end
+  end
+
+  option "with-avx2", "Build MMseqs2 with AVX2 instruction set support"
```
Shorten message by removing `MMseqs2`

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,71 @@
+    args = *std_cmake_args
+
+    args << (build.with?("mpi") ? "-DHAVE_MPI=1" : "-DHAVE_MPI=0")
+    args << (build.with?("test") ? "-DHAVE_TESTS=1" : "-DHAVE_TESTS=0")
```
Make it always `-DHAVE_TESTS=0`

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,71 @@
+    args << (build.with?("mpi") ? "-DHAVE_MPI=1" : "-DHAVE_MPI=0")
+    args << (build.with?("test") ? "-DHAVE_TESTS=1" : "-DHAVE_TESTS=0")
+
+    if build.bottle?
```
Make it always `-DHAVE_SSE4_1=1`

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,71 @@
+      args << "-DHAVE_SSE4_1=1" if build.with?("sse41")
+    end
+
+    system "cmake", ".", *args, "--debug-output", "--trace"
```
Remove `--debug-output` and `--trace`

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,71 @@
+    system "cmake", ".", *args, "--debug-output", "--trace"
+    system "make"
+
+    bin.install "src/mmseqs"
```
Please use `make install` rather than manually installing stuff.

milot-mirdita replied:
Currently `make install` also install the bash-completion script to a non-standard location. I can think about a better way to install it with the next release. Can I leave the manual install in for now?

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,71 @@
+    else
+      s += "SSE4.1 ✘. MMseqs2 will not work.\n"
+    end
+    s += "\nIf supported, AVX2 CPU instructions can be used:\n"
```
Remove the AVX2 caveats.

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,71 @@
+  end
+
+  def caveats
+    s = "MMseqs2 requires at least SSE4.1 CPU instruction support:\n"
```
Issue a caveat only if SSE4.1 is not present, to warn the user that the produced binary may not work.

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,71 @@
+
+  test do
+    system "#{bin}/mmseqs", "createdb", "#{doc}/examples/QUERY.fasta", "q"
+    system "#{bin}/mmseqs", "cluster", "q", "res", "tmp", "-s", "1", "--cascaded"
```
Can we match the output produced?

milot-mirdita replied:
With the next MMseqs2 release we will be able to. This version has a multicore instability issue, where we do not necessarily produce the same number of clusters with increasing number of threads.
I will include a check in the formula with the next release, is that okay?

fxcoudert replied:
OK

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,71 @@
+      url "https://github.com/soedinglab/MMseqs2.wiki.git", :using => :git
+    end
+  end
+
```
You also need: `bottle :disable, "Requires SSE4.1 support"`

milot-mirdita replied:
Is it necessary to disable bottle building? The compiler does not need to have SSE support.

milot-mirdita commented:
Hi @fxcoudert,

Thank you for your feedback. I incorporated most of your reviews and added a few comments, where I would be glad for more feedback. 

If everything is fine, I have to squash my commits and force push it again, right?

Best regards,
Milot

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,64 @@
+  option "with-avx2", "Build with AVX2 instruction set support"
+
+  needs :openmp
+  depends_on :mpi => [:cxx, :optional]
```
Reorder needs/depends_on so we have a single block with cmake first (build dep), :arch next, :mpi then and :openmp at the end

milot-mirdita replied:
Reordered the section.
Brew audit now tells me to use depends_on gcc, instead of needs openmp, which I have done like suggested.

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,64 @@
+  end
+
+  def install
+    inreplace "src/version/Version.cpp", /.+/m, "const char *version = \"#{version}\";"
```
Works for me. Please add a comment, and consider reporting upstream so that the tarball can be built directly.

milot-mirdita replied:
I added a comment and also improved the MMseqs2 cmake script in commit f7991db. With the next MMseqs release I will be able to pass a cmake flag instead of this hack.

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,64 @@
+  def install
+    inreplace "src/version/Version.cpp", /.+/m, "const char *version = \"#{version}\";"
+
+    args = *std_cmake_args
```
Combine two lines into `args = std_cmake_args << "-DHAVE_TESTS=0"`
No blank lines between all the `args <<`

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,64 @@
+    if build.with?("avx2")
+      args << "-DHAVE_AVX2=1"
+    else
+      args << "-DHAVE_SSE4_1=1"
```
If SSE4.1 is required, do we really have to pass it explicitly?

milot-mirdita replied:
I want to pass it explicitly, since MMseqs2 will try out by itself what instructions are available. Two brew users might thus have a differently compiled version.

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,64 @@
+    args << (build.with?("mpi") ? "-DHAVE_MPI=1" : "-DHAVE_MPI=0")
+
+    if build.with?("avx2")
+      args << "-DHAVE_AVX2=1"
```
If we have AVX2, we don't need to specify SSE4?

fxcoudert commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,64 @@
+    system "cmake", ".", *args
+    system "make"
+
+    bin.install "src/mmseqs"
```
Is there no `make install`?

milot-mirdita replied:
I replaced it with make install. However I don't really like it since my make install also install the bash completion script to a non-standard directory named `util`.

ilovezfs commented:
@BrewTestBot test this please

ilovezfs commented:
@BrewTestBot test this please

ilovezfs commented:
Unfortunately, if this cannot be bottled, we'll need to pass on it in homebrew/core but you may want to host it in [your own tap](https://github.com/Homebrew/brew/blob/master/docs/How-to-Create-and-Maintain-a-Tap.md#how-to-create-and-maintain-a-tap). Thanks for the PR nonetheless @milot-mirdita!

milot-mirdita commented:
@ilovezfs, This could be bottled. I added the `bottle :disable` due to the request of fxcoudert. Even if the build server would not have SSE4.1 support, it can compile it. However since pretty much every Intel CPU in the last decade has SSE4.1, no build server should have issues with neither compiling, nor running the test.

ilovezfs commented:
@milot-mirdita bottling will pass `-march=core2` to the build.

milot-mirdita commented:
We enable the relevant -msse flag in cmake. Also there is CPUID check that will give the user a sensible error message, if they call the binary on a CPU without SSE4.1 support (instead of some random invalid instruction during computation).

ilovezfs commented:
I willing to give it a try if you think the bottles will be OK.

ilovezfs commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,63 @@
+class Mmseqs2 < Formula
+  desc "Software suite for very fast protein sequence search and clustering"
+  homepage "https://mmseqs.org"
```
trailing slash please

ilovezfs commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,63 @@
+  version "1-c7a89"
+  sha256 "e756a0e5cb3aa8e1e5a5b834a58ae955d9594be1806f0f32800427c55f3a45d5"
+
+  head do
```
let's remove head

ilovezfs commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,63 @@
+
+  bottle :disable, "Requires SSE4.1 support"
+
+  option "with-avx2", "Build with AVX2 instruction set support"
```
this formula should have no options

milot-mirdita replied:
I would like to keep this option. If AVX2 support is available and enabled, the user gets 1.3-1.5x more throughput out of MMseqs2.

milot-mirdita commented:
Should I remove the `bottle :disable` from the formula?

ilovezfs commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,63 @@
+  depends_on "cmake" => :build
+  depends_on "gcc" => :build
+  depends_on :arch => :x86_64
+  depends_on :mpi => [:cxx, :optional]
```
decide if you want this in or out but the formula should have no options. also this has been deprecated in favor of just `depends_on "gcc"`

milot-mirdita replied:
Could you clarify please what I should change here?

You mean changing `depends_on "gcc" => :build` to `depends_on "gcc"`?

I would also like to keep optional the MPI dependency/option.

ilovezfs replied:
Right `depends_on "gcc"`

>I would also like to keep optional the MPI dependency/option.

The formula should have no options so turn it on by default :)

milot-mirdita replied:
There are a few downsides to having MPI support enabled by default:
1) We do some additional (unnecessary for single node) merge steps of the result files
3) Some CLI flags for a low-overhead mode are only available in non-mpi mode
2) Non-mpi is better tested by us

ilovezfs commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,63 @@
+    end
+  end
+
+  bottle :disable, "Requires SSE4.1 support"
```
this should be removed

ilovezfs commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,63 @@
+    system "cmake", ".", *args
+    system "make", "install"
+
+    doc.install "README.md"
```
this will be installed in `prefix` automatically

ilovezfs commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,63 @@
+
+    doc.install "README.md"
+    doc.install resource("documentation")
+    doc.install "examples"
```
these go in `pkgshare`

milot-mirdita commented:
My question probably disappeared among the last review.

I would like to keep the --with-avx2 option, since it can increase the throughput of MMseqs2 by about 1.3x-1.5x. 

I removed the MPI option.

Is that okay? The avx2 option is now the only remaining one.

fxcoudert commented:
The avx2 option will not be available for users that install a prebuilt bottle, anyway, since those need to support every possible hardware.

I suggest removing the AVX2 option, and enabling it automatically for source builds if the hardware is there:

```
args << "-DHAVE_AVX2=1" if ENV.compiler == :clang && Hardware::CPU.avx2? && !build.bottle?
```

(see in `fftw` for example)

ilovezfs commented:
For --build-from-source we pass -march=native. Does that interact correctly with what you're saying?

milot-mirdita commented:
Yes, i just tried it myself. It works correctly.

But, I just found a different issue:
```
$ brew install mmseqs2.rb --build-from-source -v
/usr/bin/sandbox-exec -f /tmp/homebrew20180112-34644-1sb0jwq.sb nice /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.3/bin/ruby -W0 -I /usr/local/Homebrew/Library/Homebrew -- /usr/local/Homebrew/Library/Homebrew/build.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/mmseqs2.rb --verbose
==> Downloading https://github.com/soedinglab/MMseqs2/archive/1-c7a89.tar.gz
Already downloaded: /Users/mirdita/Library/Caches/Homebrew/mmseqs2-1-c7a89.tar.gz
==> Verifying mmseqs2-1-c7a89.tar.gz checksum
tar xzf /Users/mirdita/Library/Caches/Homebrew/mmseqs2-1-c7a89.tar.gz
==> cmake . -DCMAKE_C_FLAGS_RELEASE=-DNDEBUG -DCMAKE_CXX_FLAGS_RELEASE=-DNDEBUG -DCMAKE_INSTALL_PREFIX=/usr/local/Cellar/mmseqs2/1-c7a89 -DCMAKE_BUILD_TYPE=Release -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_VERBOSE_MAKEFILE=ON -Wno-dev -DHAVE_TESTS=0 -DHAVE_MPI=0
-- The CXX compiler identification is AppleClang 9.0.0.9000039
```

Its not actually using GCC. I also grepped for another formula that uses cmake and gcc and found `opencoarrays`, that one is also not using gcc but the Apple Clang.

Should I submit a new ticket for the issue?

ilovezfs commented:
That is intentional. `gcc` is a shim. Does this actually fail to build with clang?

milot-mirdita commented:
No, but it won't have any multithreading, since Apple Clang does not provide OpenMP support.

milot-mirdita commented:
Should I replace the section with:

``
  depends_on "gcc"
  depends_on :arch => :x86_64

  fails_with :clang do
    cause "MMseqs2 requires OpenMP support"
  end
```

fxcoudert commented:
`needs :openmp`

milot-mirdita commented:
I had that originally, but `brew audit` was complaining about that. Can I ignore the complaint?

fxcoudert commented:
No apparently this got deprecated while I was looking elsewhere :)

ilovezfs commented:
It still needs `depends_on "gcc"`.

milot-mirdita commented:
I updated the formula to force using gcc now. Could you please take another look?
Should I squash all my changes into a single commit, assuming the build bot does not find any other issues?

ilovezfs commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,46 @@
+  desc "Software suite for very fast protein sequence search and clustering"
+  homepage "https://mmseqs.org/"
+  url "https://github.com/soedinglab/MMseqs2/archive/1-c7a89.tar.gz"
+  version "1-c7a89"
```
Any chance of requesting that this use semver?

milot-mirdita commented:
Sorry, we are keeping this schema for now. We will try to keep the CLI as stable as possible, but we do not provide a stable/public API, so semver does not really make sense.

ilovezfs commented:
@milot-mirdita semver supports that case, and would be`0.y.z`. See https://semver.org/ item 4.

milot-mirdita commented:
I mean, we will not provide an API at all. We consider the CLI as our public facing API.

We decided against using semver already a while ago during internal discussions. Reaching this compromise was already not really easy.

ilovezfs commented:
Here's the sort of thing that can happen:
```
irb(main):001:0> Version.new("1-c7a89") > Version.new("1.1-abcde")
=> false
```

milot-mirdita commented:
Only the first number will be incremented (https://github.com/soedinglab/MMseqs2/issues/46#issuecomment-340253375). So that won't happen.

ilovezfs commented:
OK. If you do decide to go with semver at some point, the equivalent semver is
```
0.0.1+c7a89
```

ilovezfs commented:
yup. 1 commit with subject `mmseqs2 1-c7a89 (new formula)`

also we standardized on `fails_with :clang # needs OpenMP support` for now

ilovezfs commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,46 @@
+  end
+
+  resource "documentation" do
+    url "https://github.com/soedinglab/MMseqs2.wiki.git", :revision => "6dbd3666edb64fc71173ee714014e88c1ebe2dfc"
```
`:revision` belongs on the next line with the `:` under the `"`

ilovezfs replied:
This URL doesn't seem to exist

ilovezfs replied:
also should we bump it to 8055a1bf70bb3923e36ba243d5c582fced4896fa ?

ilovezfs commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,46 @@
+  end
+
+  def install
+    # version information is read from git, next MMseqs2 version will include a cmake flag so we do not need this hack
```
wrap comment at 80th column

ilovezfs commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,46 @@
+    system "cmake", ".", *args
+    system "make", "install"
+
+    pkgshare.install resource("documentation")
```
probably should be `doc.install` but I can't see the contents since that URL doesn't seem to exist

ilovezfs commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,46 @@
+
+  depends_on "cmake" => :build
+  depends_on "gcc"
+  depends_on :arch => :x86_64
```
I doubt anyone is going to be trying 32 bit at this point

milot-mirdita commented:
The URL is the git url of the standard github wiki: https://github.com/soedinglab/mmseqs2/wiki

Every github repository has a wiki repository if you append .wiki.git to its git url.

ilovezfs commented:
>Every github repository has a wiki repository if you append .wiki.git to its git url. It does not have its own web interface.

TIL.

milot-mirdita commented:
:)

Should I remove the arch dependency?

ilovezfs commented:
>Should I remove the arch dependency?

I would. We can always put it back if someone is crazy enough to try that …

milot-mirdita commented:
I am mostly worried about linuxbrew (which I use a lot).

ilovezfs commented:
>linuxbrew

I have not heard of such a thing.

milot-mirdita commented:
Uhm.. I removed the arch flag and squashed everything. Is everything okay now?

ilovezfs commented:
```
Warning: mmseqs2 dependency gmp was built with a different C++ standard
library (libc++ from clang). This may cause problems at runtime.
```
Is this a concern in the case of MMseqs2?

milot-mirdita commented:
We do not use any GMP stuff, that is probably more of an issue for the gcc formula itself.

ilovezfs commented:
OK. We should add `cxxstdlib_check :skip` to this formula then so users don't ask us that question :)

ilovezfs commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,44 @@
+  sha256 "e756a0e5cb3aa8e1e5a5b834a58ae955d9594be1806f0f32800427c55f3a45d5"
+
+  depends_on "cmake" => :build
+  depends_on "gcc"
```
should be a blank line here

milot-mirdita commented:
I added the blank and the cxx check. Is it in the right order?

ilovezfs commented:
@milot-mirdita yup, and it needs a blank line under the `depends_on "gcc"`, and the commit subject should be downcased.

ilovezfs commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,46 @@
+
+  resource "documentation" do
+    url "https://github.com/soedinglab/MMseqs2.wiki.git",
+        :revision => "6dbd3666edb64fc71173ee714014e88c1ebe2dfc"
```
should we bump this to 8055a1bf70bb3923e36ba243d5c582fced4896fa?

milot-mirdita replied:
With the next release, there is already some new stuff there.

ilovezfs commented on Formula/mmseqs2.rb:
```diff
@@ -0,0 +1,46 @@
+    system "cmake", ".", *args
+    system "make", "install"
+
+    doc.install resource("documentation")
```
This is also installing the entire `.git` directory. I think we want
```
    resource("documentation").stage do
      doc.install Dir["*"]
    end

    pkgshare.install "examples"
```
to avoid that.

ilovezfs replied:
or even
```
    resource("documentation").stage { doc.install Dir["*"] }
    pkgshare.install "examples"
```

ilovezfs commented:
Thanks for the :new: formula and your first contribution to Homebrew, @milot-mirdita! Without people like you submitting PRs we couldn't run this project. You rock!
